Softeq.XToolkit.PushNotifications.iOS/Services/IosPushNotificationService.cs

```diff
@@ -0,0 +1,111 @@
+            IIosPushNotificationsConsumer pushNotificationsConsumer,
+            ILogManager logManager)
+        {
+            _pushNotificationsConsumer = pushNotificationsConsumer;
+            _logger = logManager.GetLogger<IosPushNotificationService>();
+
+            UNUserNotificationCenter.Current.Delegate = new UserNotificationCenterDelegate(pushNotificationsConsumer, logManager);
+            UNUserNotificationCenter.Current.SetNotificationCategories(new NSSet<UNNotificationCategory>(pushNotificationsConsumer.GetCategories().ToArray()));
+        }
+
+        /// <inheritdoc />
+        public async Task RegisterAsync()
+        {
+            var requiredAuthorizationOptions = _pushNotificationsConsumer.GetRequiredAuthorizationOptions();
+            var (isGranted, error) = await UNUserNotificationCenter.Current
+                .RequestAuthorizationAsync(requiredAuthorizationOptions)
+                .ConfigureAwait(false);
+
+            if (error != null)
+            {
+                throw new NSErrorException(error);
+            }
+
+            _pushNotificationsConsumer.OnPushNotificationAuthorizationResult(isGranted);
+
+            // We should register token in any case (even when permissions are not granted) to handle the possibility
+            // of user changing permission in Settings (the system itself will disregard notifications if permissions are not granted)
+            UIApplication.SharedApplication.InvokeOnMainThread(() =>
+            {
+                UIApplication.SharedApplication.RegisterForRemoteNotifications();
+            });
+        }
+
+        /// <inheritdoc />
+        public async Task UnregisterAsync()
+        {
+            UIApplication.SharedApplication.InvokeOnMainThread(() =>
+            {
+                UIApplication.SharedApplication.UnregisterForRemoteNotifications();
+            });
+
+            await _pushNotificationsConsumer
+                .OnUnregisterFromPushNotifications()
+                .ConfigureAwait(false);
+        }
+
+        /// <inheritdoc />
+        public void ClearAllNotifications()
+        {
+            UIApplication.SharedApplication.ApplicationIconBadgeNumber = 0;
+            UNUserNotificationCenter.Current.RemoveAllDeliveredNotifications();
+            UNUserNotificationCenter.Current.RemoveAllPendingNotificationRequests();
+        }
+
+        /// <inheritdoc />
+        public void RegisteredForRemoteNotifications(UIApplication application, NSData deviceToken)
+        {
+            _pushNotificationsConsumer.OnRegisteredForRemoteNotifications(application, deviceToken);
+        }
+
+        /// <inheritdoc />
+        public void FailedToRegisterForRemoteNotifications(UIApplication application, NSError error)
+        {
+            _pushNotificationsConsumer.OnFailedToRegisterForRemoteNotifications(application, error);
+        }
+
+        /// <inheritdoc />
+        public void DidReceiveRemoteNotification(
```

Comment:
Export this method to a separate class can simplify registration:
```cs
builder.Singleton<IosPushNotificationService, IPushNotificationsService>();
builder.Singleton<IosPushNotificationAppDelegate, IPushNotificationAppDelegate>();
```
instead of:
```cs
builder.Singleton<IosPushNotificationService>();
builder.Singleton<IPushNotificationsService>(container => container.Resolve<IosPushNotificationService>());
builder.Singleton<IPushNotificationAppDelegate>(container => container.Resolve<IosPushNotificationService>());
```

Reply:
Done